[DA] Fix zero coeff bug in Strong SIV test with runtime assumptions (#149991)

sebpop opened the pull request:
Fix GitHub issue #149991 where Strong SIV test incorrectly concludes 'none!' for symbolic coefficients that could be zero, leading to 0/0 undef behavior.

The Strong SIV test was incorrectly concluding "no dependence" when the
coefficient is symbolic and the delta (difference between source and
destination) is zero.

When delta=0, the Strong SIV test divides delta/coeff to get the distance.
The bug occurs when coeff is an unknown symbolic value: if coeff=0 at runtime,
then 0/0 is undefined and all iterations access the same memory location,
creating a true dependence that was being missed.

sebpop commented on llvm/lib/Analysis/DependenceAnalysis.cpp:
```diff
@@ -1282,7 +1282,28 @@ bool DependenceInfo::strongSIVtest(const SCEV *Coeff, const SCEV *SrcConst,
+      const SCEV *Zero = SE->getZero(Coeff->getType());
+
+      // Ask SCEV's range analysis if it can prove Coeff > Zero
+      if (SE->isKnownPredicate(ICmpInst::ICMP_SGT, Coeff, Zero)) {
```
The check above is `!SE->isKnownNonZero(Coeff)`, so if we are here executing this condition, it is because at compile time it is not possible to know that `Coeff != 0`.
At this point we have two choices, either try another test at compile-time `Coeff > Zero` (as you said, this may be redundant with the above call to `Coeff != 0`, although from my understanding, `SE->isKnownPredicate` is using a different computation than `SE->isKnownNonZero`) or record a runtime predicate.

As Coeff is the stride of the AddRec and both source and dest strides are equal, we can have either Coeff either positive or negative. Restricting Coeff to be positive is overly conservative.
I replaced this code with testing `Coeff != 0`.

Meinersbur commented on llvm/lib/Analysis/DependenceAnalysis.cpp:
```diff
@@ -1282,7 +1282,40 @@ bool DependenceInfo::strongSIVtest(const SCEV *Coeff, const SCEV *SrcConst,
+      const SCEV *Zero = SE->getZero(Coeff->getType());
+
+      // Ask SCEV's range analysis if it can prove Coeff != Zero.
+      if (SE->isKnownPredicate(ICmpInst::ICMP_NE, Coeff, Zero)) {
```
This is the same test as on line 1288. If they are not equally powerful (that would be some TODO to fix):
```
boo lsKnownNonZero = SE->isKnownNonZero(Coeff) || SE->isKnownPredicate(ICmpInst::ICMP_NE, Coeff, Zero);
```
There is also `DependenceInfo::isKnownPredicate`

Meinersbur commented on llvm/lib/Analysis/DependenceAnalysis.cpp:
```diff
@@ -1282,7 +1282,40 @@ bool DependenceInfo::strongSIVtest(const SCEV *Coeff, const SCEV *SrcConst,
+              Assumptions.getPredicates());
+          NewPreds.push_back(Pred);
+          const_cast<DependenceInfo *>(this)->Assumptions =
+              SCEVUnionPredicate(NewPreds, *SE);
```
Consider refactoring this. There already is `SCEVUnionPredicate::add` but it is private, mayb add a method that reurns the union.

sebpop committed:
[DA] Fix zero coefficient bug in Strong SIV test using runtime assumptions

The Strong SIV test was incorrectly concluding "no dependence" when the
coefficient is symbolic and the delta (difference between source and
destination) is zero.

When delta=0, the Strong SIV test divides delta/coeff to get the distance.
The bug occurs when coeff is an unknown symbolic value: if coeff=0 at runtime,
then 0/0 is undefined and all iterations access the same memory location,
creating a true dependence that was being missed.

sjoerdmeijer reviewed (approved):
LGTM, provided that @Meinersbur previous minor comments have been addressed, which I think is the case.

sebpop committed:
[SCEV] Add SCEVUnionPredicate::getUnionWith method

sebpop committed:
small other fixes

kasuga-fj commented on llvm/lib/Analysis/DependenceAnalysis.cpp:
```diff
@@ -1368,7 +1369,35 @@ bool DependenceInfo::strongSIVtest(const SCEV *Coeff, const SCEV *SrcConst,
+    // Check if coefficient could be zero. If so, 0/0 is undefined and we
+    // cannot conclude that only same-iteration dependencies exist.
+    // When coeff=0, all iterations access the same location.
+    if (isa<SCEVUnknown>(Coeff)) {
```
Why only `SCEVUnknown`? What happens if `Coeff` is like `%m * %n`? I think this may also be zero.

sebpop replied:
I rewrote the code to avoid enumerating all possible expressions that Coeff can take at this location:
```
    if (isa<SCEVConstant>(Coeff) || SE->isKnownNonZero(Coeff)) {
      LLVM_DEBUG(
          dbgs() << "\t    Coefficient proven non-zero by SCEV analysis\n");
    } else {
      // Cannot prove at compile time, would need runtime assumption.
[...]
    }
```

kasuga-fj commented on llvm/lib/Analysis/DependenceAnalysis.cpp:
```diff
@@ -3664,14 +3699,18 @@ DependenceInfo::depends(Instruction *Src, Instruction *Dst,
-    if (AllEqual)
+    bool AddedAssumptionsDuringTests =
+        Result.Assumptions.getPredicates().size() > AssumptionsBeforeTests;
+    if (AllEqual && !AddedAssumptionsDuringTests)
```
I think the condition should be `Result.Assumptions.getPredicates().empty()`, not `!AddedAssumptionsDuringTests`. If I understand correctly, current implementation silently discards the assumptions added before tests.

sebpop commented on llvm/lib/Analysis/DependenceAnalysis.cpp:
```diff
@@ -1368,7 +1369,35 @@ bool DependenceInfo::strongSIVtest(const SCEV *Coeff, const SCEV *SrcConst,
+      // Use both isKnownNonZero and range analysis to prove coefficient != 0.
+      bool CoeffKnownNonZero = SE->isKnownNonZero(Coeff) ||
+                               isKnownPredicate(CmpInst::ICMP_NE, Coeff,
+                                                SE->getZero(Coeff->getType()));
```
Yes, I verified that DependenceInfo::isKnownPredicate calls in all cases SE->isKnownNonZero and so it is redundant.  I removed the call to isKnownPredicate.

sebpop committed:
address comments from Ryotaro Kasuga

sebpop committed:
Address review comment: preserve all runtime assumptions, not just those added during tests

ensure we preserve all runtime assumptions, including those added before DA
tests from delinearization.
